src/common/json-schema-input/object-input.js

```diff
@@ -0,0 +1,150 @@
+import _ from 'messages'
+import React, { Component, cloneElement } from 'react'
+import forEach from 'lodash/forEach'
+import includes from 'lodash/includes'
+import map from 'lodash/map'
+
+import {
+  autobind,
+  propTypes
+} from 'utils'
+
+import { descriptionRender } from './helpers'
+
+import GenericInput from './generic-input'
+
+// ===================================================================
+
+class ObjectItem extends Component {
+  get value () {
+    return this.refs.input.value
+  }
+
+  set value (value) {
+    this.refs.input.value = value
+  }
+
+  render () {
+    const { props } = this
+
+    return (
+      <div className='p-b-1'>
+        {cloneElement(props.children, {
+          ref: 'input'
+        })}
+      </div>
+    )
+  }
+}
+
+// ===================================================================
+
+@propTypes({
+  depth: propTypes.number,
+  label: propTypes.string.isRequired,
+  required: propTypes.bool,
+  schema: propTypes.object.isRequired,
+  uiSchema: propTypes.object,
+  value: propTypes.object
+})
+export default class ObjectInput extends Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      use: Boolean(props.required),
+      children: this._makeChildren(props)
+    }
+  }
+
+  get value () {
+    const obj = {}
+
+    if (this.state.use) {
+      forEach(this.refs, (instance, key) => {
+        obj[key] = instance.value
+      })
+    }
+
+    return obj
+  }
+
+  set value (value) {
+    forEach(this.refs, (instance) => {
+      instance.value = value
+    })
+  }
+
+  @autobind
+  _handleOptionalChange (event) {
+    const { checked } = event.target
+
+    this.setState({
+      use: checked
+    })
+  }
+
+  _makeChildren (props) {
+    const {
+      depth = 0,
+      schema,
+      uiSchema = {},
+      value = {}
+    } = props
+    const obj = {}
+
+    forEach(schema.properties, (childSchema, key) => {
+      obj[key] = (
+        <ObjectItem key={key}>
+          <GenericInput
+            depth={depth + 2}
+            label={childSchema.title || key}
+            required={includes(schema.required, key)}
+            schema={childSchema}
+            uiSchema={uiSchema.properties}
+            value={value[key]}
+          />
+        </ObjectItem>
+      )
+    })
+
+    return obj
+  }
+
+  componentWillReceiveProps (props) {
```

Comment:
You should have test here if the props you are interested in have really changed.